EFCore.CheckConstraints/CheckConstraints/Internal/DiscriminatorCheckConstraintConvention.cs

```diff
@@ -0,0 +1,57 @@
+using System.Linq;
+using System.Text;
+using Microsoft.EntityFrameworkCore.Metadata.Builders;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.EntityFrameworkCore.Metadata.Conventions;
+using Microsoft.EntityFrameworkCore.Storage;
+
+namespace EFCore.CheckConstraints.Internal
+{
+    /// <summary>
+    ///     A convention that creates check constraints ensuring that (complete) discriminator columns only have
+    ///     expected values.
+    /// </summary>
+    public class DiscriminatorCheckConstraintConvention : IModelFinalizingConvention
+    {
+        readonly ISqlGenerationHelper _sqlGenerationHelper;
+
+        public DiscriminatorCheckConstraintConvention(ISqlGenerationHelper sqlGenerationHelper)
+            => _sqlGenerationHelper = sqlGenerationHelper;
+
+        /// <inheritdoc />
+        public virtual void ProcessModelFinalizing(IConventionModelBuilder modelBuilder, IConventionContext<IConventionModelBuilder> context)
+        {
+            var sql = new StringBuilder();
+
+            foreach (var (rootEntityType, discriminatorValues) in modelBuilder.Metadata
+                .GetEntityTypes()
+                .GroupBy(e => e.GetRootType())
+                .Where(g => g.Key.GetDiscriminatorProperty() != null && g.Key.GetIsDiscriminatorMappingComplete())
+                .Select(g => (g.Key, g.Select(e => e.GetDiscriminatorValue()))))
```

Comment:
LINQ is fancy but performance could be bad! :trollface:

Reply:
Indeed :) Though this runs only once at model finalization, and doesn't seem that inefficient to me, do you think there's an actual problem? We could also wait for someone to complain first to find out.

Reply:
Nobody will complain about this specifically as it's an opaque part of model building